### PR TITLE
Add OpenJDK 16ea6 Loom Early Access

### PR DESCRIPTION
**Pull request by soberich**
see https://docs.gradle.org/nightly/userguide/toolchains.html#sec:auto_detection

### CODE REVIEW

File: src/main/scala/io/sdkman/changelogs/java/OpenJdkMigrations.scala [hunk 1]
```diff
@@ -188,4 +188,31 @@ class OpenJdkMigrations {
...
+  def migrate077(implicit db: MongoDatabase): Unit =
+    Map(
+      Linux64    -> "openjdk-16-loom+6-105_linux-x64_bin.tar.gz",
+      MacOSX     -> "openjdk-16-loom+6-105_osx-x64_bin.tar.gz",
+      Windows    -> "openjdk-16-loom+6-105_windows-x64_bin.zip"
+    ).map {
+        case (platform, binary) =>
+          Version(
+            "java",
+            "16.0.6-loom", //has to be "1.2.3" e.i. numeric in order to Gradle auto detection work https://docs.gradle.org/nightly/userguide/toolchains.html#sec:auto_detection
```
Comment (marc0der) [inline src/main/scala/io/sdkman/changelogs/java/OpenJdkMigrations.scala:206]: This suffix is not currently supported on our backend yet.

Comment (eddumelendez) [reply src/main/scala/io/sdkman/changelogs/java/OpenJdkMigrations.scala:206]: what about `16.ea.6.loom-open`?

Comment (soberich) [reply src/main/scala/io/sdkman/changelogs/java/OpenJdkMigrations.scala:206]: @eddumelendez "has to be `\d\.\d\.\d.+` at the moment.
@marc0der Could you please specify what exactly do you mean?

**Comment by marc0der** [comment]
My main concern is that this is actually an ea build, and we are designating it as final release so that another tool like Gradle can detect it. I think it is better to make Gradle more flexible in its recognition.

**Comment by soberich** [comment]
It is better, but until then this is a workaround to make this possible.

**Comment by marc0der** [comment]
I'm happy for us to add loom, but then it should be called `16.ea.6.lm-open`. If that is a problem for Gradle, please do raise it with their team. They are a very responsive and friendly bunch.

**Comment by eddumelendez** [comment]
I agree with @marc0der, we can add loom as any other openjdk candidate.

**Comment by soberich** [comment]
> I'm happy for us to add loom, but then it should be called `16.ea.6.lm-open`. If that is a problem for Gradle, please do raise it with their team. They are a very responsive and friendly bunch.

Done.

File: src/main/scala/io/sdkman/changelogs/java/OpenJdkMigrations.scala [hunk 2]
```diff
@@ -188,4 +188,31 @@ class OpenJdkMigrations {
       .foreach { version =>
         removeVersion("java", "16.ea.15-open", version.platform)
       }
+
+  @ChangeSet(
+    order = "077",
```
Comment (eddumelendez) [inline src/main/scala/io/sdkman/changelogs/java/OpenJdkMigrations.scala:193]:
```suggestion
    order = "078",
```

File: src/main/scala/io/sdkman/changelogs/java/OpenJdkMigrations.scala [hunk 3]
```diff
@@ -188,4 +188,31 @@ class OpenJdkMigrations {
...
+    id = "077-add_openjdk_java_16-loom-6",
```
Comment (eddumelendez) [inline src/main/scala/io/sdkman/changelogs/java/OpenJdkMigrations.scala:194]:
```suggestion
    id = "078-add_openjdk_java_16-loom-6",
```

File: src/main/scala/io/sdkman/changelogs/java/OpenJdkMigrations.scala [hunk 4]
```diff
@@ -188,4 +188,31 @@ class OpenJdkMigrations {
...
+    author = "soberich"
+  )
+  def migrate077(implicit db: MongoDatabase): Unit =
```
Comment (eddumelendez) [inline src/main/scala/io/sdkman/changelogs/java/OpenJdkMigrations.scala:197]:
```suggestion
  def migrate078(implicit db: MongoDatabase): Unit =
```

File: src/main/scala/io/sdkman/changelogs/java/OpenJdkMigrations.scala [hunk 5]
```diff
@@ -188,4 +188,31 @@ class OpenJdkMigrations {
...
+    Map(
+      Linux64    -> "openjdk-16-loom+6-105_linux-x64_bin.tar.gz",
+      MacOSX     -> "openjdk-16-loom+6-105_osx-x64_bin.tar.gz",
+      Windows    -> "openjdk-16-loom+6-105_windows-x64_bin.zip"
+    ).map {
+        case (platform, binary) =>
+          Version(
+            "java",
+            "16.0.6.lm-open", //has to be "1.2.3" e.i. numeric in order to Gradle auto detection work https://docs.gradle.org/nightly/userguide/toolchains.html#sec:auto_detection
```
Comment (eddumelendez) [inline src/main/scala/io/sdkman/changelogs/java/OpenJdkMigrations.scala:206]:
```suggestion
            "16.0.6.lm-open",
```

File: src/main/scala/io/sdkman/changelogs/java/OpenJdkMigrations.scala [hunk 6]
```diff
@@ -188,4 +188,31 @@ class OpenJdkMigrations {
...
+            s"https://download.java.net/java/early_access/loom/6/$binary",
+            platform,
+            Some(OpenJDK)
+          )
+      }
+      .toList
+      .validate()
+      .insert()
+      .foreach { version =>
+        removeVersion("java", "16.0.6.lm-open", version.platform)
```
Comment (eddumelendez) [inline src/main/scala/io/sdkman/changelogs/java/OpenJdkMigrations.scala:216]: this is not needed

**Review by eddumelendez — CHANGES_REQUESTED** [review]
@soberich thanks for your contribution. I have made some suggestion due to changelog `077` already exists

File: src/main/scala/io/sdkman/changelogs/java/OpenJdkMigrations.scala (hunk 5 above)
Comment (marc0der) [reply src/main/scala/io/sdkman/changelogs/java/OpenJdkMigrations.scala:206]: Mind if we change this to `16.ea.6.lm-open` as we previously discussed?

Comment (eddumelendez) [reply src/main/scala/io/sdkman/changelogs/java/OpenJdkMigrations.scala:206]: suggesting that change too. goo catch @marc0der !

**Review by eddumelendez — APPROVED** [review]

**Comment by eddumelendez** [comment]
@soberich thanks for your contribution! this is now in master

**Comment by soberich** [comment]
@eddumelendez we've got a problem. `16.ea.6.lm-open` is already 15 symbols. when jdk build will have build number of two digits (which it already is), it is too long for sdkman .
`16.ea.14.lm-open` = 16 symbols

**Comment by eddumelendez** [comment]
> 16.ea.6.lm-open is already 15 symbols

it is 17 characters now

**Pull request merged**